File: testsuites/HLT/Benchmark/concurrency-benchmark/ProducerConsumer/loom/SynchProdCons.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2026. All rights reserved.
 * This source file is part of the Cangjie project, licensed under Apache-2.0
 * with Runtime Library Exception.
 * 
 * See https://cangjie-lang.cn/pages/LICENSE for license information.
 */

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicInteger;

public class SynchProdCons {

    static int nPairs = 1;
    static int nIter = 20000;

    static class Pair {
        AtomicInteger monIdx = new AtomicInteger(-1);
        AtomicInteger role = new AtomicInteger(0);
        Object[] monitors = {new Object(), new Object()};

        void test(int iter) throws Exception {
            int idx = monIdx.incrementAndGet();

            Object o1 = monitors[idx];
            Object o2 = monitors[idx ^ 1];

            synchronized (o1) {
                boolean consumer = (role.incrementAndGet() == 1); //came first - be a producer

                if (consumer) {
                    for (int cnt = 0; cnt < iter; cnt++) {
                        o1.wait();
                        synchronized (o2) { o2.notify(); }
                    }
                } else {
                    for (int cnt = 0; cnt < iter; cnt++) {
                        synchronized (o2) { o2.notify(); }
                        o1.wait();
                    }
                }
            }
        }
    }


/////////////////////////


    static void parseArgs(String args[]) throws Exception {
        for (int pos = 0; pos < args.length; pos +=2 ) {
            String option = args[pos];
            int value = Integer.parseInt(args[pos + 1]);

            if (value <= 0)
                throw new Exception();

            if (option.equals("-iter")) {
                nIter = value;
            } else if (option.equals("-threadPairs")) {
                nPairs = value;
            } else
                throw new Exception();
        }
    }

    static public void main(String args[]) throws Exception {

        try {
            parseArgs(args);
        } catch (Exception e) {
            System.out.println("Usage: [-iter <num>] [-threadPairs <numx>]");
            System.out.println("");
            System.exit(1);
        }

        System.out.println("Started: " + nIter + " iterations for " + nPairs + " thread pairs");

        new  WorkerGroup(20000, 2).execute();  // warm-up for JITs

        long elapsedTime = new WorkerGroup(nIter, nPairs).execute();

        System.out.println("Time, ms: " + elapsedTime);
    }
}


/////////////////////////


class WorkerGroup {

    private CountDownLatch startLatch;
    private Thread[] workers;

    WorkerGroup(int nIter, int nPairs) {
        startLatch = new CountDownLatch(1);

        workers = new Thread[nPairs * 2];

        int iterForPair = nIter / nPairs;

        for (int i = 0; i < nPairs; i++) {
            SynchProdCons.Pair pair = new SynchProdCons.Pair();
            workers[i * 2] =  Thread.ofVirtual().start(new Worker(iterForPair, pair));
            workers[i * 2 + 1] = Thread.ofVirtual().start(new Worker(iterForPair, pair));
        }
    }


    long execute() throws InterruptedException {
        long startTime = System.currentTimeMillis();

        startLatch.countDown();

        for (Thread w : workers) w.join();

        return (System.currentTimeMillis() - startTime);
    }


    class Worker implements Runnable {
        private int iterations;
        private SynchProdCons.Pair pair;

        Worker(int iterations, SynchProdCons.Pair pair) {
            this.iterations = iterations;
            this.pair = pair;
        }

        public void run() {
            try {
                startLatch.await();
                pair.test(iterations);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }
}

File: testsuites/HLT/Benchmark/concurrency-benchmark/ProducerConsumer/go/SynchProdCons.go
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2026. All rights reserved.
 * This source file is part of the Cangjie project, licensed under Apache-2.0
 * with Runtime Library Exception.
 * 
 * See https://cangjie-lang.cn/pages/LICENSE for license information.
 */

package main

import (
	"fmt"
	"os"
	"runtime"
	"strconv"
	"sync"
	"sync/atomic"
	"time"
)

var nPairs = 1
var nIter = 20000

var LockOSThread = false

type Pair struct {
  monIdx int64
  role   int64
  co     [2]*sync.Cond
}

func newPair(monIdx int, role int) *Pair {
  m := [2]sync.Mutex{}
  c := [2]*sync.Cond{sync.NewCond(&m[0]), sync.NewCond(&m[1])}
  return &Pair{
    monIdx: int64(monIdx),
    role:   int64(role),
    co:     c,
  }
}

func (pair *Pair) test(iter int) {

  idx := atomic.AddInt64(&(pair.monIdx), 1)

  o1 := pair.co[idx]
  o2 := pair.co[idx^1]

  o1.L.Lock()
  consumer := atomic.AddInt64(&(pair.role), 1) == 1

  if consumer {
    for cnt := 0; cnt < iter; cnt++ {
      o1.Wait()
      o2.L.Lock()
      o2.Signal()
      o2.L.Unlock()
    }
  } else {
    for cnt := 0; cnt < iter; cnt++ {
      o2.L.Lock()
      o2.Signal()
      o2.L.Unlock()
      o1.Wait()
    }
  }
  o1.L.Unlock()
}

func work(iter int, pair *Pair, startLatch, finish *sync.WaitGroup) {
  if LockOSThread {
    runtime.LockOSThread()
  }
  startLatch.Wait()
  pair.test(iter)
  finish.Done()
}

func workGroupExecute(nIter int, nPairs int) int64 {
  var startLatch, finish sync.WaitGroup
  startLatch.Add(1)
  finish.Add(nPairs * 2)

  iterForPair := nIter / nPairs

  for i := 0; i < nPairs; i++ {
    pair := newPair(-1, 0)
    go work(iterForPair, pair, &startLatch, &finish)
    go work(iterForPair, pair, &startLatch, &finish)
  }

  startTime := time.Now()
  startLatch.Done()
  finish.Wait()
  return time.Since(startTime).Milliseconds()
}

func parseArgs(args []string) {
  defer func() {
    if err := recover(); err != nil {
      fmt.Println("Usage: [-iter <num>] [-threadPairs <numx>] [-os-thread]")
      os.Exit(1)
    }
  }()
  for pos := 0; pos < len(args); pos += 2 {
    option := args[pos]

    if option == "-os-thread" {
      LockOSThread = true
      pos--
      continue
    }

    value, err := strconv.Atoi(args[pos+1])

    if value <= 0 || err != nil {
      panic("value not valid!")
    }

    if option == "-iter" {
      nIter = value
    } else if option == "-threadPairs" {
      nPairs = value
    } else {
      panic("option not valid!")
    }
  }
}

func main() {

  // fmt.Println("GOMAXPROCS:", runtime.GOMAXPROCS(0))

  args := os.Args
  parseArgs(args[1:])
  fmt.Print("Started: ", nIter, "iterations for ", nPairs, " thread pairs")
  fmt.Print(" (GOMAXPROCS = ", runtime.GOMAXPROCS(0), ", LockOSThread = ", LockOSThread, ")")
  fmt.Println()

  workGroupExecute(20000, 2)
  elapsedTiem := workGroupExecute(nIter, nPairs)

  fmt.Println("Time, ms:", elapsedTiem)
}
